src/H5FDmpio.c

```diff
@@ -2473,8 +2473,13 @@ H5FD__mpio_write_vector(H5FD_t *_file, hid_t H5_ATTR_UNUSED dxpl_id, uint32_t co
             HGOTO_ERROR(H5E_VFL, H5E_CANTGET, FAIL, "can't build MPI datatypes for I/O");
 
         /* Compute max address written to */
-        if (count > 0)
-            max_addr = s_addrs[count - 1] + (haddr_t)(s_sizes[count - 1]);
+        /* Bug: need to account for fixed size optimization */
+        if (count > 0) {
```

Comment:
A few things here:
- If count == 1 this reads off the end of the s_sizes array (maybe just change the condition to count > 1)
- The sizes can actually end anywhere in the array, not just at index 1. I'd recommend checking the last element to see if it's 0, then scanning from index 1 and up until you see a 0. So maybe something like:
`if (count > 1) {
    uint32_t idx = count - 1;
    if (s_sizes[idx] == 0)
        for (idx = 0; idx < count - 1; idx++)
            if (s_sizes[idx + 1] == 0)
                 break;
    max_addr = s_addrs[count - 1] + (haddr_t)(s_sizes[idx])
}`

Reply:
Also the comment shouldn't proclaim it as a bug, just explain what it's doing

Reply:
Actually my code is wrong too, I'll try to fix it later tonight. We might need to change the H5FD__mpio_vector_build_types function

Reply:
I think <s_sizes> returned from H5FD__mpio_vector_build_types() can be (1) or (2) as follows: 

1. <count> entries of sizes
2. Two entries for same size optimization: entry 0 is the size and entry 1 is 0
Maybe the better way is to change H5FD__mpio_vector_build_types().

Reply:
It can be any number of entries between 1 (with a 0 terminator if count > 1) and  count

Reply:
If we have time it would be great to add a test for this case. We would have to use the public H5FD interface

Reply:
Applied the patch from Neil.  
I will work on adding a test for this case.